### PR TITLE
Use webc's compat layer in more places

### PR DESCRIPTION
**Pull request by Michael-F-Bryan**
This PR moves the `wasmer_wasix::wapm` module over to the `webc` crate's compatibility layer so we are decoupled from the version of WEBC a particular file uses.

### CODE REVIEW

File: lib/wasi/src/state/env.rs [hunk 1]
```diff
@@ -804,28 +804,36 @@ impl WasiEnv {
...
 
                     // Add all the commands as binaries in the bin folder
+
+                    // FIXME(Michael-F-Bryan): We need to find a way to share
+                    // the atom bytes here (e.g. via some sort of
+                    // Arc<dyn AsRef<[u8]>> or using webc::compat::SharedBytes).
+                    // Using command.atom().to_vec() will create a copy of the
+                    // bytes and we don't want a simple Arc<[u8]> because that
+                    // means we can't use bytes from a memory-mapped file.
+
```
Comment (syrusakbary) [inline lib/wasi/src/state/env.rs:814]: Have we tried using `Bytes`? (those are cheap to clone I believe)

File: lib/wasi/src/wapm/mod.rs [hunk 2]
```diff
@@ -503,3 +438,190 @@ fn count_file_system(fs: &dyn FileSystem, path: &Path) -> u64 {
...
     total
 }
+
+#[cfg(test)]
+mod tests {
+    use std::collections::BTreeMap;
+
+    use super::*;
+
+    const PYTHON: &[u8] = include_bytes!("../../../c-api/examples/assets/python-0.1.0.wasmer");
```
Comment (theduke) [inline lib/wasi/src/wapm/mod.rs:448]: Can we add a test for one of the currently used static webserver webcs, to make sure our current hacky solution of falling back to the first atom in a dependency continues to work?

File: lib/wasi/src/state/env.rs [hunk 3]
```diff
@@ -804,28 +804,36 @@ impl WasiEnv {
...
                     if !commands.is_empty() {
                         let _ = root_fs.create_dir(Path::new("/bin"));
                         for command in commands.iter() {
-                            let path = format!("/bin/{}", command.name);
+                            let path = format!("/bin/{}", command.name());
                             let path = Path::new(path.as_str());
                             if let Err(err) = root_fs
                                 .new_open_options_ext()
-                                .insert_ro_file(path, command.atom.clone())
+                                .insert_ro_file(path, command.atom().to_vec().into())
```
[the remark this answers is not in the document]
Comment (theduke) [reply lib/wasi/src/state/env.rs:823]: I'm in favor of keeping the unsafe / ownership hack in place for now.

It's not pretty, but it's the most viable solution without changes to memfs.

File: lib/wasi/src/wapm/mod.rs [hunk 4]
```diff
@@ -503,3 +438,190 @@ fn count_file_system(fs: &dyn FileSystem, path: &Path) -> u64 {
...
+    const COREUTILS: &[u8] = include_bytes!("../../../../tests/integration/cli/tests/webc/coreutils-1.0.14-076508e5-e704-463f-b467-f3d9658fc907.webc");
+    const BASH: &[u8] = include_bytes!("../../../../tests/integration/cli/tests/webc/bash-1.0.12-0103d733-1afb-4a56-b0ef-0e124139e996.webc");
+
+    #[test]
```
Comment (theduke) [inline lib/wasi/src/wapm/mod.rs:452]: We should also have a test that makes sure the current static webserver webcs continue to work with the changes.

(they should, but wouldn't want to break this)

They currently rely on commands from a dependency being automatically used, like:

```
[package]
name = 'christoph/tmp'
version = '0.1.0'
description = 'christoph/tmp website'

[dependencies]
"sharrattj/static-web-server" = '1'

[fs]
public = 'public'
```

**Review by theduke — CHANGES_REQUESTED** [review]

File: lib/wasi/src/state/env.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Michael-F-Bryan) [reply lib/wasi/src/state/env.rs]: I've added it back in, but reduced the scope of the `unsafe` code.

File: lib/wasi/src/state/env.rs [hunk 5]
```diff
@@ -804,28 +804,36 @@ impl WasiEnv {
...
+                                    command.name(),
                                     err
                                 );
                                 continue;
                             }
 
                             // Add the binary package to the bin factory (zero copy the atom)
                             let mut package = package.clone();
-                            package.entry = Some(command.atom.clone());
+                            package.entry = Some(command.atom().to_vec().into());
```
[the remark this answers is not in the document]
Comment (Michael-F-Bryan) [reply lib/wasi/src/state/env.rs]: Added back in.

[the remark this answers is not in the document]
Comment (Michael-F-Bryan) [reply lib/wasi/src/state/env.rs]: Yup. Fixed.

File: lib/wasi/src/wapm/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Michael-F-Bryan) [reply lib/wasi/src/wapm/mod.rs:448]: Good call. I've added an assertion to the Python test.

File: lib/wasi/src/wapm/mod.rs (hunk 4 above)
Comment (Michael-F-Bryan) [reply lib/wasi/src/wapm/mod.rs:453]: I've added a test based on https://wapm.dev/wasmer/hello and it passes. Can you have a look and see if I need to make more assertions?

**Review by syrusakbary — APPROVED** [review]
LGTM

File: lib/wasi/src/wapm/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (john-sharratt) [reply lib/wasi/src/wapm/mod.rs:448]: nice

**Review by theduke — APPROVED** [review]
Note: we could store `Arc<BinaryPackage>` references on `WasiEnv` to improve the ownership safety.  (with the module cache handing out `Arc<BinaryPackage>` instead of cloning.

But that can happen inn future work.

**Pull request merged**